File: moderate/172_card_number_validation.py
import sys

def is_luhn_valid(card_number):
    digits = [int(d) for d in card_number]
    checksum = sum(digits[-1::-2])  # sum of odd digits

    for d in digits[-2::-2]:        # even digits
        checksum += sum(int(x) for x in str(d*2))
    return checksum % 10 == 0

for line in open(sys.argv[1]):
    print(int(is_luhn_valid(line.strip().replace(' ', ''))))

File: moderate/172_card_number_validation.c
#include <stdio.h>
#include <string.h>

#define LINE_SIZE   1024
#define MAX_PAN     20

int sum_of_digits(int n) {
    int sum= 0;
    while (n) {
        sum += n % 10;
        n /= 10;
    }
    return sum;
}

int is_luhn_valid(char *pan){
    int checksum = 0;
    int odd = 1;

    for (int i=strlen(pan)-1; i>=0; i--) {
        int digit = pan[i] - '0';
        checksum += odd ? digit : sum_of_digits(digit * 2);
        odd ^= 1;
    }

    return (checksum % 10 == 0);
}


int main(int argc, const char * argv[]) {
    FILE *file = fopen(argv[1], "r");
    char line[LINE_SIZE];
    char pan[MAX_PAN];

    while (fgets(line, LINE_SIZE, file)) {
        char *p;
        int len_pan = 0;
        for (p=line; *p; p++)
            if ('0' <= *p && *p <= '9')
                pan[len_pan++] = *p;
        pan[len_pan] = '\0';

        printf("%d\n", is_luhn_valid(pan));
    }

    return 0;
}
